include/gatekeeper_acl.h

```diff
@@ -25,16 +25,17 @@
 
 struct acl_search {
 	/* References to the start of the IP header in each packet. */
-	const uint8_t   *data[GATEKEEPER_MAX_PKT_BURST];
+	const uint8_t   **data;
 	/* References to each packet's mbuf. */
-	struct rte_mbuf *mbufs[GATEKEEPER_MAX_PKT_BURST];
+	struct rte_mbuf **mbufs;
 	/* The classification results for each packet. */
-	uint32_t        res[GATEKEEPER_MAX_PKT_BURST];
+	uint32_t        *res;
 	/* The number of packets held for classification. */
 	unsigned int    num;
```

Comment:
After looking at the ACL code, I found that if we store the ``acl_search_entry`` in a list at the end of the ``struct acl_search``, then the ``data`` and ``res`` fields in each entry will be segmented from each other. However, function ``int rte_acl_classify (const struct rte_acl_ctx *ctx, const uint8_t **data, uint32_t *results, uint32_t num, uint32_t categories)`` will be called in ``process_acl()`` to classify the packets' data in an array. Thus, we may have to call ``rte_acl_classify()`` for each ``acl_search_entry``, which will degrade the classification performance. Maybe, we can allocate the memory as a whole, and still keep each of the fields as an array?

Reply:
You are right.

I think field `res` is only used in `process_acl()`, so we can allocate field `res` as a local variable of `process_acl()`. With that, `struct acl_search` would be as follows:
```
struct acl_search {
	/* Fixed field here. */
	struct rte_mbuf *mbufs;
	const uint8_t   *data[0];
};
```
During initialization, not only you need to calculate the needed size of the memory, but also set field `mbufs` as `mbufs = (struct rte_mbuf *)data[x];` for a suitable `x`.

Reply:
Thanks! The field ``res`` indeed can be declared as a local variable of ``process_acl()``.

By looking at function ``add_pkt_acl(struct acl_search *acl, struct rte_mbuf *pkt)``, shouldn't ``data[x]`` point to the start of the IP header of ``mbufs[x]`` for each ``x in [0, num - 1]``?

Reply:
I see your point. We can allocate memory after ``data[0]``, then we can have a point ``mbufs``, which points to the start of the memory region storing the real packets.

Reply:
When I tried to do using the suggested solution, there is an error ``error: cast discards ‘const’ qualifier from pointer target type [-Werror=cast-qual]``, since we are adding ``const`` to the ``data[0]`` field. Indeed, the ``const`` is necessary for this ``data[0]`` field, as it will be used by ``rte_acl_classify()``, which requires the ``const`` qualifier.